### PR TITLE
Allow portable agent to run multiple test assemblies

### LINKED ISSUE #13: Add ability to run multiple assemblies
Agent should manage running multiple assemblies and combining the results, to avoid this functionality being duplicated in each runner.

### PR DESCRIPTION
**Pull request by ChrisMaddock**
This is a checkpoint PR, for some feedback on a few points before tidying up for merge. (Fixes #13)

@rprouse, @CharliePoole - I'm going to put some specific questions in line comments, but would appreciate feedback on the general approach - this isn't a part of NUnit I'm too familiar with! The intention is to run assemblies in serial for now.

**Still to-do**
- [x] Decide on behaviour of Load methods
- [x] Additional per-method tests (current test is just to check it was actually working!)
- [x] Decide on behaviour of async running

### CODE REVIEW

File: src/NUnit.Portable.Agent/NUnitPortableDriver.cs [hunk 1]
```diff
@@ -69,17 +65,15 @@ public class NUnitPortableDriver
...
 
-            _frameworkControllerType = _frameworkController.GetType();
+            var testPackage = new TestPackage(testAssembly, frameworkAssembly, frameworkController);
+            _testPackages.Add(testPackage);
 
-            log.Info("Loading {0} - see separate log file", _testAssembly.FullName);
-            return ExecuteMethod(LOAD_METHOD) as string;
+            log.Info("Loading {0} - see separate log file", testAssembly.FullName);
+            return testPackage.ExecuteMethod(LOAD_METHOD) as string;
         }
```
Comment (ChrisMaddock) [inline src/NUnit.Portable.Agent/NUnitPortableDriver.cs:77]: The intention was that runners will still call the agent's `Load()` method per assembly. In this case, what should `Load()` return, the XML result from the single assembly you've just loaded, or XML for all the assemblies loaded so far?

I had a look around - the result from this isn't used in dotnet-test-nunit, and the console abstracts at a different level, so it loads all the assemblies together, and then returns a single XML result. What would a result from `Load()` actually be used for? I'm not sure of the best way to go here.

File: src/NUnit.Portable.Agent/NUnitPortableDriver.cs [hunk 2]
```diff
@@ -114,18 +119,23 @@ public string Run(Action<string> callback, string filter)
...
-            log.Info("Running {0} - see separate log file", _testAssembly.FullName);
-            ExecuteMethod(RUN_ASYNC_METHOD, new[] { typeof(Action<string>), typeof(string) }, callback, filter);
+            CheckAssembliesLoaded();
+
+            foreach (var package in _testPackages)
+            {
+                log.Info("Running {0} - see separate log file", package.FullName);
+                package.ExecuteMethod(RUN_ASYNC_METHOD, new[] { typeof(Action<string>), typeof(string) }, callback, filter);
+            }
         }
```
Comment (ChrisMaddock) [inline src/NUnit.Portable.Agent/NUnitPortableDriver.cs:130]: I'm not entirely clear what happens with the `RUN_ASYNC_METHOD` in the framework - would the way this is implemented here effectively run assemblies in parallel? I'd like to avoid that - is there a way to get around that already built-in, of would the agent need to hijack the callback somehow?

File: test/NUnit.Portable.Agent.Tests/NUnitPortableDriverTests.cs [hunk 3]
```diff
@@ -122,7 +114,7 @@ public void RunTestsAction_AfterLoad_ReturnsRunnableSuite()
             _driver.Load(_frameworkAssembly, _mockAssembly, _settings);
             var result = XmlHelper.CreateXElement(_driver.Run(null, EMPTY_FILTER));
```
Comment (ChrisMaddock) [inline test/NUnit.Portable.Agent.Tests/NUnitPortableDriverTests.cs]: @rprouse - I didn't initially realise that wrapping the results up how I currently have done would change the XML, and return a test-run rather than a test-suite. Would this cause any problems for dotnet-test-nunit, should it be done differently?

File: src/NUnit.Portable.Agent/NUnitPortableDriver.cs (hunk 2 above)
Comment (ChrisMaddock) [reply src/NUnit.Portable.Agent/NUnitPortableDriver.cs:130]: Now I've typed out the words 'hijack the callback', that seems like almost definitely the solution...

**Comment by CharliePoole** [comment]
@ChrisMaddock This is a part of NUnit I'm entirely unfamiliar with myself! I'm going to use the PR to try to start learning.

File: src/NUnit.Portable.Agent/NUnitPortableDriver.cs (hunk 1 above)
Comment (CharliePoole) [reply src/NUnit.Portable.Agent/NUnitPortableDriver.cs:77]: In the framework, load returns only the top level Xml node for the assembly. That's because the result may be remoted and not all runners need more than the top level. Runners that need more call Explore, which gets the entire content. In the console, that load result is used because it may contain error messages. I think you need to do that as soon as you start loading multiple assemblies unless you want the entire run to fail when one assembly cannot be found.

File: src/NUnit.Portable.Agent/TestPackage.cs [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    internal class TestPackage
+    {
+        internal Assembly TestAssembly { get; }
+        internal Assembly FrameworkAssembly { get; }
+        internal object FrameworkController { get; }
+
+        private Type FrameworkControllerType => FrameworkController.GetType();
+        public string FullName => TestAssembly.FullName;
+
+        internal TestPackage(Assembly testAssembly, Assembly frameworkAssembly, object frameworkController)
```
Comment (CharliePoole) [inline src/NUnit.Portable.Agent/TestPackage.cs]: TestPackage is a fundamental class in the engine API, representing a test file and a number of associated settings.

Changing it to mean something else in this agent doesn't seem like such a good idea, particularly as we try to get netstandard assemblies to run under the engine. There are a wide range of meaningful names available at relatively low cost. :-)

**Review by CharliePoole — COMMENTED** [review]
Neither approving nor disapproving... just commenting. :-)

File: src/NUnit.Portable.Agent/NUnitPortableDriver.cs (hunk 1 above)
Comment (ChrisMaddock) [reply src/NUnit.Portable.Agent/NUnitPortableDriver.cs:77]: Aha - did not know that! In that case, returning just the xml node for the assembly you are loading sounds preferable, and the runner can resolve errors. Thanks @CharliePoole!

File: src/NUnit.Portable.Agent/TestPackage.cs (hunk 4 above)
Comment (ChrisMaddock) [reply src/NUnit.Portable.Agent/TestPackage.cs]: Sure - sounds sensible!

**Comment by ChrisMaddock** [comment]
@CharliePoole, @rprouse, from my point of view, this is now ready for review/merge.

Particular 'points of note':
- I'd missed initially that the async-run callback produced progress messages rather than just normal results. Accordingly, I presume the correct behaviour is just to pass the runner-specified call back through. Is the implemention for RunAsync correct here?
- @rprouse - This PR changes the xml returned from the portable agent to return `<test-runs>` rather than `<test-suites>`. Will that cause a problem for `dotnet-test-nunit`?

Thanks both!

File: test/NUnit.Portable.Agent.Tests/NUnitPortableDriverTests.cs [hunk 5]
```diff
@@ -49,14 +50,6 @@ public void CreateDriver()
...
         }
 
-        #region Construction Test
-
-        public void ConstructController_MissingFile_ThrowsArgumentInvalid()
-        {
-            Assert.That(new NUnitPortableDriver(), Throws.ArgumentException);
-        }
-        #endregion
-
```
Comment (ChrisMaddock) [inline test/NUnit.Portable.Agent.Tests/NUnitPortableDriverTests.cs:59]: I presume this is old code? The method wasn't run as a test and doesn't seem relevant to the current implementation - I removed it.

File: test/NUnit.Portable.Agent.Tests/NUnitPortableDriverTests.cs [hunk 6]
```diff
@@ -144,19 +140,5 @@ public void RunTestsAction_WithoutLoad_ThrowsInvalidOperationException()
...
-            public Action<string> Action { get; private set; }
-
-            public CallbackEventHandler()
-            {
-                Action = s => Result = s;
-            }
-
-            public string Result { get; private set; }
-        }
-        #endregion
```
Comment (ChrisMaddock) [inline test/NUnit.Portable.Agent.Tests/NUnitPortableDriverTests.cs:160]: Was also unused

**Comment by CharliePoole** [comment]
@ChrisMaddock My own comments are just "observations" since I don't know enough about this agent to do more than that... it's down to @rprouse.

**Comment by rprouse** [comment]
@ChrisMaddock did you test these changes with the adapter? To test the combination, I create a branch of dotnet-test-nunit and switch it to use a CI build of this project then test. Did you do that?

**Comment by ChrisMaddock** [comment]
@rprouse - I haven't, I just ran the in-solution tests. Will try it tomorrow.

**Comment by ChrisMaddock** [comment]
@rprouse - Sorry, busy week! Just tried it now, it seems to pass all tests. Afraid I don't have anything in the way of core projects to test it against further.

It looks like dotnet-test-nunit actually uses the ResultSummary class from the agent to sumarise test results, so my concerns about whether it would understand `test-suite` are _probably_ unfounded, unless I've missed another potential failure point.

**Comment by rprouse** [comment]
Thanks. I will see if I can do some additional tests then merge. I am
hoping to get a release out this weekend.

**Comment by rprouse** [comment]
I've used this PR in the dotnet-test-project. Looks good so far. Merging.
